File: templates/ts-llm/src/agent.ts
// TypeScript LLM agent template.
//
// Platform contract:
//   * validated input JSON arrives via the AGENT_INPUTS env var
//     (cluster path) or on stdin (local-dev fallback)
//   * the final output JSON is the last line on stdout
//   * logs go to stderr (captured into Loki, viewable in the UI)
//   * the env var AGENT_TRACE_ID, if set, ties logs to a distributed trace

import Anthropic from "@anthropic-ai/sdk";

const MODEL = process.env.MODEL ?? "claude-sonnet-4-6";

function log(msg: string): void {
  process.stderr.write(msg + "\n");
}

async function readInputs(): Promise<Record<string, unknown>> {
  const fromEnv = (process.env.AGENT_INPUTS ?? "").trim();
  if (fromEnv) return JSON.parse(fromEnv);
  const chunks: Buffer[] = [];
  for await (const c of process.stdin) chunks.push(c as Buffer);
  const raw = Buffer.concat(chunks).toString("utf8").trim();
  return raw ? JSON.parse(raw) : {};
}

async function main(): Promise<void> {
  const inputs = await readInputs();
  const question = String(inputs.question ?? "Say hello.");

  log(`trace=${process.env.AGENT_TRACE_ID ?? ""} model=${MODEL} question=${JSON.stringify(question)}`);

  const client = new Anthropic();
  const resp = await client.messages.create({
    model: MODEL,
    max_tokens: 512,
    messages: [{ role: "user", content: question }],
  });

  const answer = resp.content
    .filter((b) => b.type === "text")
    .map((b) => (b as { type: "text"; text: string }).text)
    .join("")
    .trim();

  process.stdout.write(JSON.stringify({ answer }) + "\n");
}

main().catch((e) => {
  log("error: " + (e as Error).stack);
  process.exit(1);
});

File: templates/python-llm/agent.py
"""Hello-LLM agent template.

The platform contract:
  - validated input JSON arrives via the AGENT_INPUTS env var (cluster path)
    or on stdin (local-dev convenience)
  - your final output JSON is the last line on stdout
  - logs go to stderr (captured into Loki, viewable in the UI)
  - the env var AGENT_TRACE_ID, if set, ties your logs to a distributed trace
"""

from __future__ import annotations

import json
import os
import sys

import anthropic


MODEL = os.environ.get("MODEL", "claude-sonnet-4-6")


def log(msg: str) -> None:
    print(msg, file=sys.stderr, flush=True)


def read_inputs() -> dict:
    raw = os.environ.get("AGENT_INPUTS", "").strip()
    if not raw:
        raw = sys.stdin.read().strip()
    return json.loads(raw) if raw else {}


def main() -> int:
    payload = read_inputs()
    question = payload.get("question", "Say hello.")

    trace_id = os.environ.get("AGENT_TRACE_ID", "")
    log(f"trace={trace_id} model={MODEL} question={question!r}")

    client = anthropic.Anthropic()
    resp = client.messages.create(
        model=MODEL,
        max_tokens=512,
        messages=[{"role": "user", "content": question}],
    )

    text_parts = [block.text for block in resp.content if getattr(block, "type", "") == "text"]
    answer = "".join(text_parts).strip()

    print(json.dumps({"answer": answer}))
    return 0


if __name__ == "__main__":
    sys.exit(main())
